File: Algorithms/7. Tree/Traversal/traversal.cpp
// C++ program for different tree traversals
#include <iostream>
using namespace std;
 
// A binary tree node
struct Node {
  int data;
  struct Node* left;
  struct Node* right;
  Node(int data) {
    this->data = data;
    left = right = NULL;
  }
};
 
// preorder: root, left, right
void printPreorder(struct Node* node) {
  if (node == NULL) // reach leave node, go to prev level
    return;

  cout << node->data << " "; // output the value
  printPreorder(node->left); // recurse on left subtree
  printPreorder(node->right); // recurse on right subtree
} 

// inorder: left, root, right
void printInorder(struct Node* node) {
  if (node == NULL) // reach leave node, go to prev level
    return;

  printInorder(node->left); // recurse on left child
  cout << node->data << " "; // print the data of node
  printInorder(node->right); // recurse on right child
}

// postorder: left, right, root
void printPostorder(struct Node* node) {
  if (node == NULL) // reach leave node, go to prev level
    return;

  printPostorder(node->left); // recurse on left subtree
  printPostorder(node->right); // recurse on right subtree
  cout << node->data << " ";  // output the value
}

// testing the functions
int main() {
  // create Nodes
  Node* root = new Node(1);
  root->left = new Node(2);
  root->right = new Node(3);
  root->left->left = new Node(4);
  root->left->right = new Node(5); 

  // output binary tree
  cout << "Output the tree's structure:\n";
  cout << "      1            \n"
       << "     / \\          \n"
       << "    2   3         \n"
       << "   / \\              \n"
       << "  4   5            \n";

  // output different traversal approaches' results
  cout << "\nPreorder:  root, left, right\n";
  printPreorder(root);
  cout << "\nInorder:   left, root, right\n";
  printInorder(root); 
  cout << "\nPostorder: left, right, root\n";
  printPostorder(root);
}

File: Algorithms/7. Tree/Traversal/traversal.py
# a binary tree node
class Node:
  def __init__(self, data):
    self.data = data
    self.left = None
    self.right = None

# print binary tree in preorder: root, left, right
def print_preorder(node):
  if node == None: # reached leaf node, go back to prev level
    return
  
  print(str(node.data) + " "), # output the current value
  print_preorder(node.left) # recurse on left subtree
  print_preorder(node.right) # recurve on right subtree

# print binary tree in inorder: left, root, right
def print_inorder(node):
  if node == None: # reached leaf node, go back to prev level
    return
  
  print_inorder(node.left) # recurse on left subtree
  print(str(node.data) + " "), # output the current value
  print_inorder(node.right) # recurse on right subtree

# print binary tree in postorder: left, right, root
def print_postorder(node):
  if node == None: # reached leaf node, go back to prev level
    return
  
  print_postorder(node.left) # recurse on left subtree
  print_postorder(node.right) # recurse on right subtree
  print(str(node.data) + " "), # output the value

# ================================
# testing the traversal approaches
# ================================
root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.right = Node(5)

# output binary tree
print("Output the tree's structure:")
print("      1   ")
print("     / \\ ")
print("    2   3 ")
print("   / \\   ")
print("  4   5   ")

# output different traversal approaches' results
print("\n- Preorder:  root, left, right")
print_preorder(root)
print("\n- Inorder:   left, root, right")
print_inorder(root)
print("\n- Postorder: left, right, root")
print_postorder(root)